### PR TITLE
Add test that proves you can write a ByteReference using its iterator

### PR DESCRIPTION
**Pull request by iverase**
New test that updates the content of a ByteReference by calling ByteReference#iterator.

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
Test LGTM. Could we make `org.elasticsearch.common.bytes.AbstractBytesReference#iterator` abstract since its (unused-in-production) default implementation does not satisfy this? Also it deserves a note in the Javadocs for `org.elasticsearch.common.bytes.BytesReference#iterator` to say that this method _must_ return direct references to the raw pages and not copies.

**Comment by iverase** [comment]
umm, there is one implementation that is problematic: `ZeroBytesReference` as it does not really contains internal pages so not sure what to do in that case.

**Comment by DaveCTurner** [comment]
`ZeroBytesReference` is only used in tests, we don't care about its writethrough behaviour so can just suppress `testWriteWithIterator` in its test suite.

**Comment by DaveCTurner** [comment]
If you were feeling diligent you could even make `org.elasticsearch.common.bytes.ZeroBytesReference#iterator` return an iterator that asserts that the page is always full of zero bytes on each call to `next()`, to help protect against someone accidentally relying on its write-through behaviour in a future test. Up to you tho.

File: server/src/main/java/org/elasticsearch/common/bytes/BytesReference.java [hunk 1]
```diff
@@ -184,7 +184,11 @@ static BytesReference fromByteArray(ByteArray byteArray, int length) {
 
     /**
      * Returns a BytesRefIterator for this BytesReference. This method allows
-     * access to the internal pages of this reference without copying them. Use with care!
+     * access to the internal pages of this reference without copying them.
+     * It must return direct references to the pages, not copies. If the object
+     * does not hold internal pages, then it mush return Use with care!
+     *
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/bytes/BytesReference.java:190]: Unfinished comment edit I think :)

File: server/src/test/java/org/elasticsearch/common/bytes/ZeroBytesReference.java [hunk 2]
```diff
@@ -56,20 +58,23 @@ public BytesRef toBytesRef() {
...
-        final byte[] buffer = new byte[PageCacheRecycler.BYTE_PAGE_SIZE];
+        final byte[] buffer = new byte[Math.min(length, PageCacheRecycler.BYTE_PAGE_SIZE)];
         return new BytesRefIterator() {
-
             int remaining = length;
+            BytesRef current;
 
             @Override
             public BytesRef next() {
+                if (current != null) {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/common/bytes/ZeroBytesReference.java:68]: Not sure we need to track `current` here, we can just look at `buffer` directly.

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I have only superficial comments remaining.

File: server/src/test/java/org/elasticsearch/common/bytes/ZeroBytesReference.java (hunk 2 above)
Comment (iverase) [reply server/src/test/java/org/elasticsearch/common/bytes/ZeroBytesReference.java]: doh! of course....

File: server/src/main/java/org/elasticsearch/common/bytes/BytesReference.java (hunk 1 above)
Comment (iverase) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReference.java]: I wanted to add something about implementations without internal pages, I removed it.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by iverase** [comment]
@elasticmachine update branch

**Comment by iverase** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Pull request merged**